### PR TITLE
Issue #3595: [UX] Add summary text for the filter fieldset

### LINKED ISSUE #3595: [UX] Add a summary text next to the CKEditor formatting options link, showing the active format
Follow-up to #3516

**Original issue title:** Add summary text for the fieldset that allows selecting a text editor

In #3516 we changed the text size and appearance of the `Formatting options` link below CKEditor. In the issue, we also discussed related improvements and it was suggested to keep making these changes later, cf. https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-461993952. I've read through the issue thread again, and here are my follow-up suggestions:

1. Add a summary text behind the formatting options link, showing the active format. See https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460044306 and https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460287284.

2. Change terminology from "Formatting options" to "Editor details": See https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-460832014 and https://github.com/backdrop/backdrop-issues/issues/3516#issuecomment-461584101.

I'm making the second suggestion because on most of my sites, for content editors the `Formatting options` aren't options at all. Usually, editors have only the "Filtered HTML" permission, so that there is no *option* in the fieldset but only *information*. With a name like `Editor details`, the link makes also sense when there are no options.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2550

### PR DESCRIPTION
**Pull request by klonos**
fixes https://github.com/backdrop/backdrop-issues/issues/3595

### CODE REVIEW

**Comment by klonos** [comment]
Note for when merging this: commit message should change to `Issue #3595: [UX] Add summary text for the filter fieldset`

**Comment by klonos** [comment]
...refreshing PR sandbox.

**Comment by klonos** [comment]
Refreshing PR sandbox...

**Comment by jenlampton** [comment]
closing for new sandbox.

**Comment by jenlampton** [comment]
Sandboxme?

File: core/modules/filter/js/filter.js
Comment (docwilmot) [inline core/modules/filter/js/filter.js]: I suspect that since this is already within the scope of `fieldset.filter-wrapper`, then `...find('select.filter-list' :selected')` should be specific enough. But havent tested.

**Comment by olafgrabienski** [comment]
One small issue: `.fieldset-title` ("Formatting options") and `summary` aren't well vertical-aligned. `.fieldset-legend span.summary` has `top: .4em;`, should probably be `.5em`, at least in this case.

**Comment by klonos** [comment]
Hmm 🤔 ...I don't see that misalignment @olafgrabienski. Can you please post a screenshot?

**Comment by klonos** [comment]
OK, I see it now.

@olafgrabienski this may be a generic issue with fieldset summaries (I haven't touched any CSS in this PR). Do you mind opening a new issue to see if we need to fix this in Seven in general?

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: Thanks @docwilmot 👍 ...I am about to push another commit, in which I also address the case when a single text format is available to the current user. I have gone with `select.filterEditors-processed` and `input.filterEditors-processed` instead, which is shorter, but uses the common `.filterEditors-processed` class.

File: core/modules/filter/filter.module
Comment (klonos) [inline core/modules/filter/filter.module]: I went with this approach here, in order to pass the human-readable name of the text format to the js. Please let me know if using something like the following is preferred instead:

```php
  backdrop_add_js(array('editor-name' => $options[$element['#format']]), 'setting');
```

**Comment by klonos** [comment]
...the single test failure in the php7 test is one of the usual random ones:

`fail "Generated 1 URL alias." found path_pattern.test:943`

😞

**Comment by olafgrabienski** [comment]
> this may be a generic issue with fieldset summaries (...) Do you mind opening a new issue to see if we need to fix this in Seven in general?

@klonos I wouldn't mind but the only example of comparable fieldset summaries that I know was removed in https://github.com/backdrop/backdrop-issues/issues/3657. Do you know of another example?

**Comment by klonos** [comment]
@olafgrabienski I cannot remember off of the top of my head any admin page that has fieldsets, but in any case, I've created https://github.com/backdrop/backdrop-issues/issues/4460 to deal with that misalignment issue specifically (perhaps you'd like to file a PR for it?).

**Comment by olafgrabienski** [comment]
> I've created backdrop/backdrop-issues#4460 to deal with that misalignment issue specifically (perhaps you'd like to file a PR for it?)

Sure, I put it on my list, but will probably wait until *this* PR is merged.

**Comment by ghost** [comment]
@klonos Care to refresh just to confirm test failure is random?

File: core/modules/filter/filter.module
Comment (ghost) [inline core/modules/filter/filter.module]: I think this could be simplified:

```suggestion
        // Add the human-readable name of the text format so it can be used as
        // the fieldset summary.
```

File: core/modules/filter/filter.module
Comment (ghost) [reply core/modules/filter/filter.module]: I think this approach is fine (it appears we do this elsewhere in core too), I'd just suggest a different name for the data attribute...

```suggestion
        'data-format-name' => array($options[$element['#format']]),
```

File: core/modules/filter/js/filter.js
Comment (ghost) [inline core/modules/filter/js/filter.js]: I don't think mentioning CKEditor here is necessary or correct...

```suggestion
 * Provides summary text for the "Editor details" fieldset, under each textarea
 * field with a text format.
```

File: core/modules/filter/js/filter.js
Comment (ghost) [inline core/modules/filter/js/filter.js]: Is this really necessary...?

File: core/modules/filter/js/filter.js
Comment (ghost) [inline core/modules/filter/js/filter.js]: I think we don't need so much description here, we can check for the existence of each element via its length() (https://stackoverflow.com/a/587408/4168619) and we can move the actual getting of the name part to the if statement (as I think this reads a bit nicer).

```suggestion
      // Look for a select list of text formats.
      var $select_list = $element.find('select.filterEditors-processed :selected');
      // Otherwise look for a hidden input element.
      var $input_element = $element.find('input.filterEditors-processed');

      if ($select_list.length) {
        $summary = $select_list.text();
      }
      else if ($input_element.length) {
        $summary = $input_element.attr('data-format-name');
      }
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: core/modules/filter/filter.module
Comment (klonos) [reply core/modules/filter/filter.module]: Agreed, but better be less vague and use `data-text-format-name`

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: Yup. The fieldset breaks otherwise. Do you have a suggestion?

Comment (ghost) [reply core/modules/filter/js/filter.js]: Interesting. Wonder why...?

I haven't tested it, but I'd have thought that just replacing `$element` with `$(element)` in this function should work...

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: Right 👍 ...but it should be "with a text editor" instead. Also, the name of the fieldset is "Formatting options":

image

Let's fix all that.

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: ...nah, scratch that; my bad. It's not needed indeed, and nothing breaks.

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: Excellent suggestion 👍 ...I would still want us to include "when the current user has access to only a single text format" ...it took me a bit to figure that out 😅

**Comment by klonos** [comment]
Tests passed this time, but the PR sandbox is broken 😞 ...refreshing.

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: ...yes it does 😞 ...had to revert this.

Comment (klonos) [reply core/modules/filter/js/filter.js]: My js-fu is not that strong @BWPanda ...happy for you to push a change with whatever works, but removing this line causes the fieldset to break for me:

image

Comment (klonos) [reply core/modules/filter/js/filter.js]: ...also, see: https://github.com/backdrop/backdrop/blob/1.x/core/misc/collapse.js#L7

File: core/modules/filter/js/filter.js
Comment (ghost) [inline core/modules/filter/js/filter.js]: @klonos It works for me. Here's what I meant:

```suggestion
      var $summary = '';
      // Look for a select list of text formats.
      var $select_list = $(element).find('select.filterEditors-processed :selected');
      // Otherwise look for a hidden input element (when the current user has
      // access to only a single text format).
      var $input_element = $(element).find('input.filterEditors-processed');
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

File: core/modules/filter/filter.module
Comment (quicksketch) [inline core/modules/filter/filter.module]: I don't think this needs to be an array here. When output, this creates `data-text-format-name=["Formatted HTML"]`. Additionally, this needs to be filtered through check_plain().

```suggestion
        'data-text-format-name' => check_plain($options[$element['#format']]),
```

File: core/modules/filter/js/filter.js
Comment (quicksketch) [inline core/modules/filter/js/filter.js]: Convention is to only put a `$` in front of variables that contain jQuery objects. This `$summary` is a string, so it should just be `summary`.

**Review by quicksketch — CHANGES_REQUESTED** [review]

**Comment by herbdool** [comment]
@klonos this looks pretty close. Do you want to make the code changes or I could apply them to get this finished.

File: core/modules/filter/js/filter.js
Comment (klonos) [reply core/modules/filter/js/filter.js]: Done: [`69d0652` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/69d065266f3651694d76fd6abc491af58b98071f)

File: core/modules/filter/filter.module
Comment (klonos) [reply core/modules/filter/filter.module]: Done: [`edb5555` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/edb5555dcd328c2318c5ad3e0ba1fbc1801cba7d)

**Comment by klonos** [comment]
All done @quicksketch and @herbdool (also rebased). Sorry it has taken me so long to get back to you on this.

File: core/themes/basis/css/component/fieldset.css
Comment (indigoxela) [inline core/themes/basis/css/component/fieldset.css]: As an alternative for switching from position absolute to relative in combination with float, you might be able to achieve the same with

```
[dir="rtl"] .fieldset-legend span.summary {
  right: auto;
  left: 2em;
}
```
Then you'd also not need the negative top margin, I suppose. Didn't actually test it, though, it just seems like a smaller change to me. Worth a try?

Comment (klonos) [reply core/themes/basis/css/component/fieldset.css]: @indigoxela and @jenlampton they way you suggest to change this is how the respective selector is set in Seven. This is Basis though, and this is what it looks like if I try your suggestion:

image

...my code fixes that, but I don't claim to be a CSS expert. Can either of you please file a PR against mine if you still think that it should be fixed differently, or suggest other CSS that might work better? Thanks.

Comment (klonos) [reply core/themes/basis/css/component/fieldset.css]: ...I fiddled a bit with the CSS, and I managed to make it work with `position: absolute;` as before. See this change: [`b81ec61` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/b81ec61a407476e6c493d691a623b469cf2192e1)

Comment (indigoxela) [reply core/themes/basis/css/component/fieldset.css]: @klonos oops, you ripped off the sandbox I was just playing with. :laughing: 

What I suggested - really only a suggestion:

```
.fieldset-legend span.summary {
  position: absolute;
  top: .7em;
  color: #999;
  right: .5em;/* LTR */
}

[dir="rtl"] .fieldset-legend span.summary {
  right: auto;
  left: .5em;
}
```
Not sure, if this is complete.

File: core/modules/system/css/system.theme.css
Comment (jenlampton) [inline core/modules/system/css/system.theme.css]: I don't think we need the AA comment (I removed mine from the other PR, they were only there as a note to myself while working on it)

File: core/themes/basis/css/component/fieldset.css
Comment (jenlampton) [inline core/themes/basis/css/component/fieldset.css]: I think this will affect all fieldset legends, which may break existing sites. Let's restore this line.

File: core/themes/basis/css/component/fieldset.css
Comment (jenlampton) [inline core/themes/basis/css/component/fieldset.css:61]: this should be enough to do what we need :)

File: core/themes/basis/css/component/fieldset.css
Comment (jenlampton) [inline core/themes/basis/css/component/fieldset.css]: remove comment

File: core/modules/system/css/system.theme.css
Comment (klonos) [reply core/modules/system/css/system.theme.css]: Removed with [`58a030d` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/58a030d3d20c84360a342697a683b48e4bdd65d8). I'm adding a comment in the PR instead.

File: core/themes/basis/css/component/fieldset.css
Comment (klonos) [reply core/themes/basis/css/component/fieldset.css]: Removed with [`58a030d` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/58a030d3d20c84360a342697a683b48e4bdd65d8). I'm adding a comment in the PR instead.

File: core/modules/system/css/system.theme.css
Comment (klonos) [inline core/modules/system/css/system.theme.css:246]: WCAG AA compliance

File: core/themes/basis/css/component/fieldset.css
Comment (klonos) [inline core/themes/basis/css/component/fieldset.css:107]: WCAG AA compliance

File: core/themes/basis/css/component/fieldset.css
Comment (klonos) [reply core/themes/basis/css/component/fieldset.css]: Done 👍🏼 ...see: [`601a916` (#2550)](https://github.com/backdrop/backdrop/pull/2550/commits/601a916a8efd2c26c1cd782ecd5947166d2dd886)

**Review by quicksketch — APPROVED** [review]

**Comment by quicksketch** [comment]
I merged into 1.x for 1.20.0. Since it comes out tomorrow I didn't see a big reason to backport to 1.19.x.

**Pull request merged**
